core/src/doc/lives.rs

```diff
@@ -205,4 +97,161 @@ impl<'a> Document<'a> {
+		sender: &Sender<Notification>,
+	) -> Result<(), Error> {
+		trace!(
+			"Called check_lqs_and_send_notifications with {} live statements",
+			live_statements.len()
+		);
+		for lv in live_statements {
+			// Create a new statement
+			let lq = Statement::from(*lv);
+			// Get the event action
+			let met = if stm.is_delete() {
+				Value::from("DELETE")
+			} else if self.is_new() {
+				Value::from("CREATE")
+			} else {
+				Value::from("UPDATE")
+			};
+			// Check if this is a delete statement
+			let doc = match stm.is_delete() {
+				true => &self.initial,
+				false => &self.current,
+			};
+			// Ensure that a session exists on the LIVE query
+			let sess = match lv.session.as_ref() {
+				Some(v) => v,
+				None => {
+					trace!("live query did not have a session, skipping");
+					continue;
+				}
+			};
+			// Ensure that auth info exists on the LIVE query
+			let auth = match lv.auth.clone() {
+				Some(v) => v,
+				None => {
+					trace!("live query did not have auth info, skipping");
+					continue;
+				}
+			};
+			// We need to create a new context which we will
+			// use for processing this LIVE query statement.
+			// This ensures that we are using the session
+			// of the user who created the LIVE query.
+			let mut lqctx = Context::background();
+			lqctx.add_value("auth", sess.pick(SD.as_ref()));
+			lqctx.add_value("scope", sess.pick(SC.as_ref()));
+			lqctx.add_value("token", sess.pick(TK.as_ref()));
+			lqctx.add_value("session", sess);
+			// We need to create a new options which we will
+			// use for processing this LIVE query statement.
+			// This ensures that we are using the auth data
+			// of the user who created the LIVE query.
+			let lqopt = opt.new_with_perms(true).with_auth(Arc::from(auth));
+			// Add $before, $after, $value, and $event params
+			// to this LIVE query so that user can use these
+			// within field projections and WHERE clauses.
+			lqctx.add_value("event", met);
+			lqctx.add_value("value", self.current.doc.deref());
+			lqctx.add_value("after", self.current.doc.deref());
+			lqctx.add_value("before", self.initial.doc.deref());
+			// First of all, let's check to see if the WHERE
+			// clause of the LIVE query is matched by this
+			// document. If it is then we can continue.
+			match self.lq_check(&lqctx, &lqopt, txn, &lq, doc).await {
+				Err(Error::Ignore) => {
+					trace!("live query did not match the where clause, skipping");
+					continue;
+				}
+				Err(e) => return Err(e),
+				Ok(_) => (),
+			}
+			// Secondly, let's check to see if any PERMISSIONS
+			// clause for this table allows this document to
+			// be viewed by the user who created this LIVE
+			// query. If it does, then we can continue.
+			match self.lq_allow(&lqctx, &lqopt, txn, &lq, doc).await {
+				Err(Error::Ignore) => {
+					trace!("live query did not have permission to view this document, skipping");
+					continue;
+				}
+				Err(e) => return Err(e),
+				Ok(_) => (),
+			}
```

Comment:
I love how documented this is, do you think we can create a module called session pathways? it could basically be a series of steps performed on the session, reducing the size of this function, and also enabling more unit testing

Reply:
Yes I had considered that. But ultimately this is the only place where we are doing it. It could be composable, but it adds a lot of overhead for what we want. But I have tried that in the past :) It's in one of the closed PRs 🤣